### ISSUE TITLE
Move block jpa implementation to thread pool

### ISSUE THREAD

**Issue by rj-hwang** [simter/simter-kv#6]

**Comment by rj-hwang** [comment]
Reference reactor official document appendix - [B.1. How do I wrap a synchronous, blocking call?](https://projectreactor.io/docs/core/release/reference/#faq.wrap-blocking),  use `Schedulers.elastic` to creates a dedicated thread to wait for the blocking resource without tying up some other resource:

```
Mono blockingWrapper = Mono.fromCallable(() -> {  // 1
  return /* make a remote synchronous call */             // 2 Return the asynchronous, blocking resource
});
blockingWrapper = blockingWrapper.subscribeOn(Schedulers.elastic()); // 3 
```

But another link [Reactive/Blocking Conversions](https://musigma.blog/2016/11/21/reactor.html#reactiveblocking-conversions), it also think about `subscribeOn` and `publishOn`. I think this is more clear way to convert block to reactive:

- a fast subscriber but slow publisher - obtaining the collection is expensive
    ```
    Flux<Person> asyncPersonLookup(PersonRepository repository) {
      return Flux.defer(() -> Flux.fromIterable(repository.findAll()))
        .subscribeOn(Schedulers.elastic());
    }
    ```
- a fast publisher and a slow subscriber - store a record into database
    ```
    Mono<Void> asyncSavePersons(Flux<Person> flux, PersonRepository repository) {
      return flux.publishOn(Schedulers.parallel())
        .doOnNext(repository::save)
        .then();
    }
    ```

**Comment by rj-hwang** [comment]
Besides, could also define a custom thread-pool size scheduler:

```
@Value("${spring.datasource.maximum-pool-size}") Integer connectionPoolSize;

Scheduler blockScheduler = Schedulers.fromExecutor(Executors.newFixedThreadPool(connectionPoolSize));

Flux.defer(() -> Flux.fromIterable(repository.findAll()))
  .subscribeOn(blockScheduler);
```

Code form [here](https://stackoverflow.com/questions/42299455/spring-webflux-and-reading-from-database#answer-44118833).

**Comment by rj-hwang** [comment]
The main one is transaction management. In JDBC, transactions are possible only within a single java.sql.Connection. To make several operations in one transaction, they have to share a connection. If we want to make some calculations in between them, we have to keep the connection.

**Comment by rj-hwang** [comment]
For a entire review, all related public method of Mono and Flux list below (From [reactor-core-3.2.6]):

```
public final Mono<T> publishOn(Scheduler scheduler)
public final Flux<T> publishOn(Scheduler scheduler)
public final Flux<T> publishOn(Scheduler scheduler, int prefetch)
public final Flux<T> publishOn(Scheduler scheduler, boolean delayError, int prefetch)

public final Mono<T> subscribeOn(Scheduler scheduler);
public final Flux<T> subscribeOn(Scheduler scheduler);
public final Flux<T> subscribeOn(Scheduler scheduler, boolean requestOnSeparateThread);
```
 
I build some test for `publishOn` and `subscribeOn`, check below code:

- [PublishOnTest.java](https://github.com/start-java/start-reactor/blob/master/src/test/java/tech/simter/start/reactor/mono/PublishOnTest.java)
- [SubscribeOnTest.java](https://github.com/start-java/start-reactor/blob/master/src/test/java/tech/simter/start/reactor/mono/SubscribeOnTest.java)

My test result:

- `publishOn`  influence all the downstream until the next occurrence of `publishOn`, include nested chain.
- `subscribeOn` influence the upstream and downstream **from the top of the chain but before any `publishOn`**. If take `subscribeOn` multiple times, only the first is valid. Take `subscribeOn` in the nested chain not influence the parent chain.


[reactor-core-3.2.6]: https://github.com/reactor/reactor-core/tree/v3.2.6.RELEASE

**Comment by rj-hwang** [comment]
Encapsulate common usage to [simter-reactive-jpa](https://github.com/simter/simter-reactive-jpa).

**Comment by rj-hwang** [comment]
Add [ReactiveDataJpaTest]  to instead of [DataJpaTest] for reactive JPA test.

This annotation inherits from `DataJpaTest` and only disabled `Transactional`. Because `EntityManager` could not share between threads, and in reactive world operation do not guarantee run on same thread all the time. So no need to start `Transactional` on the unit test class level.

[ReactiveDataJpaTest] was encapsulate to [simter-reactive-test].

[ReactiveDataJpaTest]: https://github.com/simter/simter-reactive-test/blob/master/src/main/java/tech/simter/reactive/test/jpa/ReactiveDataJpaTest.java
[DataJpaTest]: https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-testing.html#boot-features-testing-spring-boot-applications-testing-autoconfigured-jpa-test
[simter-reactive-test]: https://github.com/simter/simter-reactive-test

**Comment by rj-hwang** [comment]
Resolved this issue by [commit-5a4b732..b7ef36c](https://github.com/simter/simter-kv/compare/5a4b732..b7ef36c)
